From 4ba534d7e8b49671b6cb733476c6b1e1a2d3568b Mon Sep 17 00:00:00 2001
From: Fred Underwood <underwoodfred@proton.me>
Date: Sat, 5 Jul 2025 07:30:30 +1000
Subject: [PATCH 1/2] fix Keyguard slice

Upstream refactor in commit 1f73b12d5dd4db6ebcad5d444c79ac18cc59e3c3
removed too much and as a result the legacy Keyguard slice code was
broken. This adds back the needed elements.
---
 .../KeyguardSliceViewControllerTest.java      |  2 +-
 .../keyguard/KeyguardSliceViewController.java | 36 ++++++++++++++++++-
 .../sections/KeyguardSliceViewSection.kt      |  6 ++++
 3 files changed, 42 insertions(+), 2 deletions(-)

diff --git a/packages/SystemUI/multivalentTests/src/com/android/keyguard/KeyguardSliceViewControllerTest.java b/packages/SystemUI/multivalentTests/src/com/android/keyguard/KeyguardSliceViewControllerTest.java
index 9d10011159c7..e6c27d4290ee 100644
--- a/packages/SystemUI/multivalentTests/src/com/android/keyguard/KeyguardSliceViewControllerTest.java
+++ b/packages/SystemUI/multivalentTests/src/com/android/keyguard/KeyguardSliceViewControllerTest.java
@@ -71,7 +71,7 @@ public class KeyguardSliceViewControllerTest extends SysuiTestCase {
         when(mView.getContext()).thenReturn(mContext);
         mController = new KeyguardSliceViewController(mHandler, mBgHandler, mView,
                 mActivityStarter, mConfigurationController, mDumpManager,
-                mDisplayTracker);
+                mDisplayTracker, null, null);
         mController.setupUri(KeyguardSliceProvider.KEYGUARD_SLICE_URI);
     }
 
diff --git a/packages/SystemUI/src/com/android/keyguard/KeyguardSliceViewController.java b/packages/SystemUI/src/com/android/keyguard/KeyguardSliceViewController.java
index a8de43332556..8567f748fe59 100644
--- a/packages/SystemUI/src/com/android/keyguard/KeyguardSliceViewController.java
+++ b/packages/SystemUI/src/com/android/keyguard/KeyguardSliceViewController.java
@@ -18,6 +18,8 @@ package com.android.keyguard;
 
 import static android.app.slice.Slice.HINT_LIST_ITEM;
 
+import static com.android.systemui.util.kotlin.JavaAdapterKt.collectFlow;
+
 import android.app.PendingIntent;
 import android.net.Uri;
 import android.os.Handler;
@@ -43,11 +45,17 @@ import com.android.systemui.dagger.qualifiers.Background;
 import com.android.systemui.dagger.qualifiers.Main;
 import com.android.systemui.dump.DumpManager;
 import com.android.systemui.keyguard.KeyguardSliceProvider;
+import com.android.systemui.keyguard.domain.interactor.KeyguardInteractor;
 import com.android.systemui.plugins.ActivityStarter;
+import com.android.systemui.power.domain.interactor.PowerInteractor;
+import com.android.systemui.power.shared.model.ScreenPowerState;
 import com.android.systemui.settings.DisplayTracker;
 import com.android.systemui.statusbar.policy.ConfigurationController;
 import com.android.systemui.util.ViewController;
 
+import kotlin.coroutines.CoroutineContext;
+import kotlin.coroutines.EmptyCoroutineContext;
+
 import java.io.PrintWriter;
 import java.util.List;
 import java.util.Map;
@@ -71,6 +79,8 @@ public class KeyguardSliceViewController extends ViewController<KeyguardSliceVie
     private Uri mKeyguardSliceUri;
     private Slice mSlice;
     private Map<View, PendingIntent> mClickActions;
+    private KeyguardInteractor mKeyguardInteractor;
+    private PowerInteractor mPowerInteractor;
 
     ConfigurationController.ConfigurationListener mConfigurationListener =
             new ConfigurationController.ConfigurationListener() {
@@ -110,7 +120,9 @@ public class KeyguardSliceViewController extends ViewController<KeyguardSliceVie
             ActivityStarter activityStarter,
             ConfigurationController configurationController,
             DumpManager dumpManager,
-            DisplayTracker displayTracker) {
+            DisplayTracker displayTracker,
+            KeyguardInteractor keyguardInteractor,
+            PowerInteractor powerInteractor) {
         super(keyguardSliceView);
         mHandler = handler;
         mBgHandler = bgHandler;
@@ -118,6 +130,28 @@ public class KeyguardSliceViewController extends ViewController<KeyguardSliceVie
         mConfigurationController = configurationController;
         mDumpManager = dumpManager;
         mDisplayTracker = displayTracker;
+        mKeyguardInteractor = keyguardInteractor;
+        mPowerInteractor = powerInteractor;
+    }
+
+    @Override
+    public void onInit() {
+        setupUri(null);
+        startCoroutines(EmptyCoroutineContext.INSTANCE);
+    }
+
+    void startCoroutines(CoroutineContext context) {
+        collectFlow(mView, mKeyguardInteractor.getDozeTimeTick(),
+                (Long millis) -> {
+                    refresh();
+                }, context);
+
+        collectFlow(mView, mPowerInteractor.getScreenPowerState(),
+                (ScreenPowerState powerState) -> {
+                    if (powerState == ScreenPowerState.SCREEN_TURNING_ON) {
+                        refresh();
+                    }
+                }, context);
     }
 
     @Override
diff --git a/packages/SystemUI/src/com/android/systemui/keyguard/ui/view/layout/sections/KeyguardSliceViewSection.kt b/packages/SystemUI/src/com/android/systemui/keyguard/ui/view/layout/sections/KeyguardSliceViewSection.kt
index 962e6089dce9..5479ed50f604 100644
--- a/packages/SystemUI/src/com/android/systemui/keyguard/ui/view/layout/sections/KeyguardSliceViewSection.kt
+++ b/packages/SystemUI/src/com/android/systemui/keyguard/ui/view/layout/sections/KeyguardSliceViewSection.kt
@@ -28,8 +28,10 @@ import com.android.systemui.customization.R as customR
 import com.android.systemui.dagger.qualifiers.Background
 import com.android.systemui.dagger.qualifiers.Main
 import com.android.systemui.dump.DumpManager
+import com.android.systemui.keyguard.domain.interactor.KeyguardInteractor
 import com.android.systemui.keyguard.shared.model.KeyguardSection
 import com.android.systemui.plugins.ActivityStarter
+import com.android.systemui.power.domain.interactor.PowerInteractor
 import com.android.systemui.res.R
 import com.android.systemui.settings.DisplayTracker
 import com.android.systemui.statusbar.lockscreen.LockscreenSmartspaceController
@@ -47,6 +49,8 @@ constructor(
     val configurationController: ConfigurationController,
     val dumpManager: DumpManager,
     val displayTracker: DisplayTracker,
+    val keyguardInteractor: KeyguardInteractor,
+    val powerInteractor: PowerInteractor,
 ) : KeyguardSection() {
     private lateinit var sliceView: KeyguardSliceView
 
@@ -70,6 +74,8 @@ constructor(
                 configurationController,
                 dumpManager,
                 displayTracker,
+                keyguardInteractor,
+                powerInteractor,
             )
         controller.init()
     }
-- 
2.43.0

